File: catkin_ws_vps_v/src /kf_mpc_controller4/src/least_square.cpp
#include "../include/kf_mpc_controller4/least_square.h"

least_square::least_square()
{

}
least_square::least_square(std::vector<float> t, std::vector<float> x, int order)
{
  _t = t;
  _x = x;
  _order = order;
}
//least_square:: ~least_square()
//{

//}
Eigen::VectorXf least_square::get_Result()
{
  Eigen::Map<Eigen::VectorXf> sampleT(_t.data(),_t.size());
  Eigen::Map<Eigen::VectorXf> sampleX(_x.data(),_x.size());
  Eigen::MatrixXf mtxVandermonde(_t.size(),_order+1);
  Eigen::VectorXf colVandermonde = sampleT;
  for (size_t i = 0; i < _order + 1; ++i)
  {
    for (size_t j = 0; j< _order +1; ++j)
    {
      mtxVandermonde(i,j) = pow(colVandermonde(i),j);
    }
  }
  Eigen::VectorXf result = (mtxVandermonde.transpose()*mtxVandermonde).inverse()*(mtxVandermonde.transpose())*sampleX;
  return result;
}
Eigen::VectorXf least_square::get_Result2()
{
  int N = _order;
  // 创建A矩阵
  Eigen::MatrixXd A(_t.size(), N + 1);

  for (unsigned int i = 0; i < _t.size(); ++i) {  // 遍历所有点

    for (int n = N, dex = 0; n >= 1; --n, ++dex) {  // 遍历N到1阶
      A(i, dex) = pow(_t[i], n);
    }

    A(i, N) = 1;  //
  }
  // 创建B矩阵
  Eigen::MatrixXd B(_x.size(), 1);

  for (unsigned int i = 0; i < _x.size(); ++i) {
    B(i, 0) = _x[i];
  }
  // 创建矩阵W
  Eigen::MatrixXd W;
  W = (A.transpose() * A).inverse() * A.transpose() * B;
  // 打印W结果
  Eigen::VectorXf ret;
  ret.resize(N+1);
  for(int i=0;i<N+1;i++)
  {
    ret[i] = W(i);
  }
  //std::cout<<"--------------ret------------"<<ret<<std::endl;
  return ret;//wn--wn-1---wn-2--->w0
}

bool least_square::set_T(std::vector<float> t)
{
  _t = t;
}

bool least_square::set_X(std::vector<float> x)
{
  _x = x;
}

bool least_square::set_Order(int order)
{
  _order =  order;
}

std::vector<float> least_square::get_T()
{
  return _t;
}

std::vector<float> least_square::get_X()
{
  return _x;
}

int least_square::get_Order()
{
  return _order;
}

File: DTUAVSIM/Script/Algorithm/Predictor/LeastSquare.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using MathNet.Numerics.LinearAlgebra.Single;
using MathNet.Numerics.LinearAlgebra.Generic;

namespace DTUAVCARS.Algorithm.Predictor
{
    public class LeastSquare
    {
        private List<float> _t;
        private List<float> _x;
        private int _order;
        private DenseMatrix _A;
        private DenseMatrix _B;
        private List<float> _ret;
        public List<float> GetT()
        {
            return _t;
        }

        public List<float> GetX()
        {
            return _x;
        }

        public int GetOrder()
        {
            return _order;
        }

        public List<float> GetResult()
        {
            _A = new DenseMatrix(_t.Count, _order + 1);
            _B = new DenseMatrix(_x.Count, 1);
            for (int i = 0;i<_t.Count;++i)
            {
                for(int n= _order, dex=0;n>=1;--n,++dex)
                {
                    _A.At(i, dex, Mathf.Pow(_t[i], n));
                }
                _A.At(i, _order, 1);
            } 
            for(int i=0;i<_x.Count;++i)
            {
                _B.At(i, 0, _x[i]);
            }
            var W = (_A.Transpose() * _A).Inverse() * _A.Transpose() * _B;
            for(int i=0;i< _order + 1;i++)
            {
                _ret[i] = W.At(i, 0);
            }
            return _ret;///wn--wn-1---wn-2--->w0
        }

        public bool SetT(List<float> t)
        {
            _t = t;
            return true;
        }

        public bool SetX(List<float> x)
        {
            _x = x;
            return true;
        }
        public bool SetOrder(int order)
        {
            _order = order;
            return true;
        }

        public LeastSquare()
        {

        }
        public LeastSquare(List<float> t, List<float> x, int order)
        {
            _order = order;
            _t = t;
            _x = x;
            _A = new DenseMatrix(_t.Count, _order + 1);
            _B = new DenseMatrix(_x.Count, 1);
            _ret = new List<float>();
            for (int i = 0; i < _order + 1; i++)
            {
                _ret.Add(0.0f);
            }
        }
    }
}
